docs/block-processing/types.md

```diff
@@ -0,0 +1,951 @@
+
+```haskell
+data SecretProof = SecretProof
+    { spExtraGen       :: Scrape.ExtraGen
+    , spProof          :: Scrape.Proof
+    , spParallelProofs :: Scrape.ParallelProofs
+    , spCommitments    :: [Scrape.Commitment]
+    }
+```
+
+### Opening, openings map
+
+*keywords: `Opening`, `Secret`, `OpeningsMap`*
+
+An opening is sent at the second stage of SSC. It reveals the participant's
+secret (which can be thought of as a simple bytestring).
+
+```haskell
+newtype Opening = Opening (AsBinary Secret)
+```
+
+An openings map (`OpeningsMap`) is a map from participants' IDs to their
+openings. It has no invariants (since the opening doesn't have a public key
+in it).
+
+```haskell
+type OpeningsMap = HashMap StakeholderId Opening
+```
+
+### Encrypted/decrypted share, shares map
+
+*keywords: `EncShare`, `DecShare`; `SharesMap`, `InnerSharesMap`*
+
+A share is a part of a secret (see Wikipedia on [secret sharing][]). With
+more than a certain percentage of decrypted shares (`DecShare`) it is
+possible to recover the secret that was used to create the shares. An
+encrypted share (`EncShare`) can be decrypted with the corresponding VSS
+keypair.
+
+[secret sharing]: https://en.wikipedia.org/wiki/Secret_sharing
+
+```haskell
+newtype DecShare = DecShare Scrape.DecryptedShare
+
+newtype EncShare = EncShare Scrape.EncryptedSi
+```
+
+A shares map (`SharesMap`) is a set of shares created by some participants
+and decrypted by other participants:
+
+  * outer key = who decrypted the share
+  * inner key = who created the share
+
+```haskell
+type SharesMap = HashMap StakeholderId InnerSharesMap
+
+type InnerSharesMap = HashMap StakeholderId (NonEmpty (AsBinary DecShare))
+```
+
+So, an `InnerSharesMap` contains all shares decrypted by a certain
+participant, and keys of `InnerSharesMap` denote who created some particular
+set of shares.
+
+## Delegation-related types
+
+Note: this section talks about heavy delegation only.
+
+### Proxy key
+
+*keywords: `ProxySecretKey`, `ProxyCert`, `ProxySKHeavy`*
+
+A proxy key (`ProxySecretKey w`) is a `PublicKey`, stored in
+`pskDelegatePk`, that comes equipped with three things:
+
+  * `pskOmega` – a value of type `w`
+
+  * `pskIssuerPk` – a key which delegated something (e.g. block issuing
+    rights) to `pskDelegatePk`
+
+  * `pskCert` – a proof that `pskIssuerPk` has delegated to `pskDelegatePk`
+
+```haskell
+data ProxySecretKey w = ProxySecretKey
+    { pskOmega      :: w              -- auxiliary value
+    , pskIssuerPk   :: PublicKey      -- delegation happens from this key
+    , pskDelegatePk :: PublicKey      --   ...to this key
+    , pskCert       :: ProxyCert w    -- a proof by 'pskIssuerPk'
+    }
+```
+
+A `ProxyCert` is just a signature of `(pskDelegatePk, pskOmega)`:
+
+```
+newtype ProxyCert w = ProxyCert { unProxyCert :: CC.XSignature }
+```
+
+An invariant of `ProxySecretKey` is that its `pskCert` must be valid. TODO:
+is this right? If yes, why not `UnsafeProxySecretKey`?
```

Comment:
AFAIK, it's wrong, this invariant is not enforced (see e. g. `Bi` instance, it's trivial). Instead we have `verifyPsk` function and hopefully call it in all places where it's necessary.

Reply:
Is there any particular reason why we don't enforce it in `Bi`?

Reply:
I don't think so. Anyway, we are currently revising our approach to checked/unchecked deserialization, unsafe constructors, type invariants, etc., so probably it will be fixed in the corresponding refactoring. But for now let's describe existing code.